packages/vite/src/node/ssr/runtime/types.ts

```diff
@@ -0,0 +1,185 @@
+    id: string,
+  ): Promise<any>
+  /**
+   * Run externalized module.
+   * @param file File URL to the external module
+   */
+  runExternalModule(file: string): Promise<any>
+}
+
+export interface ModuleCache {
+  promise?: Promise<any>
+  exports?: any
+  evaluated?: boolean
+  map?: DecodedMap
+  meta?: FetchResult
+  /**
+   * Module ids that imports this module
+   */
+  importers?: Set<string>
+  imports?: Set<string>
+}
+
+export type FetchResult = ExternalFetchResult | ViteFetchResult
+
+export interface ExternalFetchResult {
+  /**
+   * The path to the externalized module starting with file://,
+   * by default this will be imported via a dynamic "import"
+   * instead of being transformed by vite and loaded with vite runtime
+   */
+  externalize: string
+  /**
+   * Type of the module. Will be used to determine if import statement is correct.
+   * For example, if Vite needs to throw an error if variable is not actually exported
+   */
+  type?: 'module' | 'commonjs' | 'builtin' | 'network'
+}
+
+export interface ViteFetchResult {
+  /**
+   * Code that will be evaluated by vite runtime
+   * by default this will be wrapped in an async function
+   */
+  code: string
+  /**
+   * File path of the module on disk.
+   * This will be resolved as import.meta.url/filename
+   */
+  file: string | null
+}
+
+export type ResolvedResult = (ExternalFetchResult | ViteFetchResult) & {
+  id: string
+}
+
+export type FetchFunction = (
+  id: string,
+  importer?: string,
+) => Promise<FetchResult>
+
+export interface ViteRuntimeOptions {
+  /**
+   * Root of the project
+   */
+  root: string
+  /**
+   * A method to get the information about the module.
+   * For SSR, Vite exposes `server.ssrFetchModule` function that you can use here.
+   * For other runtime use cases, Vite also exposes `fetchModule` from its main entry point.
+   */
+  fetchModule: FetchFunction
+  /**
+   * Custom environment variables available on `import.meta.env`. This doesn't modify the actual `process.env`.
+   */
+  environmentVariables?: Record<string, any>
+  /**
+   * Configure how source maps are resolved. Prefers `node` if `process.setSourceMapsEnabled` is available.
+   * Otherwise it will use `prepareStackTrace` by default.
+   */
+  sourcemapInterceptor?:
+    | false
+    | 'node'
+    | 'prepareStackTrace'
+    | InterceptorOptions
+  /**
+   * Disable HMR or configure HMR options.
+   */
+  hmr?:
+    | false
+    | {
+        /**
+         * Configure how HMR communicates between the client and the server.
+         */
+        connection: HMRRuntimeConnection
+        /**
+         * Configure HMR logger.
+         */
+        logger?: false | HMRLogger
+      }
+  /**
+   * Custom module cache. If not provided, creates a separate module cache for each ViteRuntime instance.
+   */
+  moduleCache?: ModuleCacheMap
+  /**
+   * Resolved modules that will be returned instead of executing the code.
+   */
+  requestStubs?: Record<string, any>
```

Comment:
Question: Could I use `fetchModule` to replace the stubs? Are stubs only replacing modules that already exists (like test stubs)? Or stubs are like virtual modules that can replace a missing module?

Reply:
Technically yes. This is a leftover from vite-node - I think we can remove it because unlike vite-node, `fetchModule` is always called in vite runtime (in vite-node it's called after `requestStubs` are checked)

Reply:
Removed now